api/certs/cert.go

```diff
@@ -0,0 +1,230 @@
+	}
+	bundle, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return false, err
+	}
+	unRevoked, err := VerifyOCSP(bundle)
+	if !unRevoked {
+		return false, err
+	}
+	certs, err := certcrypto.ParsePEMBundle(bundle)
+	if err != nil {
+		return false, err
+	}
+	if certs[0].Subject.CommonName != domain {
+		return false, fmt.Errorf("inconsistent domain names")
+	}
+	return true, nil
+}
+
+func VerifyLocalCertificate(certPath string, keyPath string, domain string) (bool, error) {
+	if len(certPath) > 0 && len(keyPath) > 0 && len(domain) > 0 && common.FileExisted(certPath) && common.FileExisted(keyPath) {
+		return VerifyCertificate(certPath, keyPath, domain)
+	}
+	return false, nil
+}
+
+func CertExists() (bool, bool, bool) {
+	wssUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpWssCert, config.Parameters.HttpWssKey, config.Parameters.HttpWssDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	httpsUserCertExists, err := VerifyLocalCertificate(config.Parameters.HttpsJsonCert, config.Parameters.HttpsJsonKey, config.Parameters.HttpsJsonDomain)
+	if err != nil {
+		log.Error(err)
+	}
+	defaultDomain, err := node.GetDefaultDomainFromIP(config.Parameters.Hostname, config.Parameters.DefaultTlsDomainTmpl)
+	if err != nil {
+		log.Error(err)
+	}
+	_, _, certFile, keyFile := GetACMEFileNames(defaultDomain, config.Parameters.CertDirectory)
+	defaultCertExists, err := VerifyLocalCertificate(certFile, keyFile, defaultDomain)
+	if err != nil {
+		log.Error(err)
+	}
+
+	return wssUserCertExists, httpsUserCertExists, defaultCertExists
+}
+
+func PrepareCerts() (chan struct{}, chan struct{}) {
+	wssCertExists, httpsCertExists, defaultCertExists := CertExists()
+	ChangeToDefault(httpsCertExists, wssCertExists)
+
+	wssCertReady := make(chan struct{}, 1)
+	httpsCertReady := make(chan struct{}, 1)
+
+	if wssCertExists || defaultCertExists {
+		go func() {
+			wssCertReady <- struct{}{}
+		}()
+	}
+	if httpsCertExists || defaultCertExists {
+		go func() {
+			httpsCertReady <- struct{}{}
+		}()
+	}
+	if !(wssCertExists && httpsCertExists) {
+		go prepareCerts(wssCertExists, httpsCertExists, wssCertReady, httpsCertReady)
+	}
+	return wssCertReady, httpsCertReady
+}
+
+func prepareCerts(wssCertExists, httpsCertExists bool, wssCertReady, httpsCertReady chan struct{}) {
+	first := true
+	for {
+		cm, err := NewCertManager()
+		if err != nil {
+			log.Errorf("create new certManager err: %v", err)
+			return
+		}
+		if !first {
+			time.Sleep(config.Parameters.CertCheckInterval * time.Second)
+		}
+		first = false
+		if cm.NeedObtain() {
+			cert, err := cm.ObtainNewCert()
+			if err != nil {
+				log.Errorf("apply cert failed: %v", err)
+				continue
+			}
+			err = saveCertToDisk(cert)
+			if err != nil {
+				log.Errorf("save cert failed: %v", err)
+				continue
+			}
+			if !wssCertExists {
+				select {
+				case wssCertReady <- struct{}{}:
+					log.Info("wss cert ready")
+				default:
+					log.Info("no wss cert available yet")
```

Comment:
This log is not right because cert is available. It should be something like wss cert channel full